0027-tcp-load-balancer.md

```diff
@@ -0,0 +1,264 @@
+        {
+            if min > num_conns
+            {
+                min_conns = num_conns;
+                min_id = Some(id);
+            }
+        }
+
+        min_id
+    }
+}
+
+```
+
+#### Upstream Health Check
+
+The health checker will simply connect to an upstream server, send a ping and receive a pong message and then disconnect.
+
+It will continue to do this indefinitely every 30 seconds (hard coded for development speed).
+
+If the upstream server rejects the connection, does not reply to the ping  within 1 second or disconnects while the connection is active, the upstream server is marked unhealthy.
+
+If the upstream completes a full connect - ping - disconnect cycle, it is again marked healthy.
+
+```rust
+enum UpstreamState
+{
+    HEALTHY,
+    UNHEALTHY,
+}
+```
+
+```rust
+enum PingState
+{
+    IDLE,
+    PING_SENT(i64),
+    PONG_RECEIVED(i64),
+}
+```
+
+```rust
+struct HealthChecker
+{
+    up_stream       : TcpStream,
+    ping_state      : PingState,
+    UpstreamState   : UpstreamState,
+    last_update     : chrono::NaiveDateTime,
+}
+```
+
+```rust
+struct ServerGroup
+{
+    ...
+    server_health   : HashMap<u32, HealthChecker>,
+}
+```
+
+
+### Server
+
+The server will listen on port 8443 for client connections, using std::net::TcpListener to listen connections from clients.
+
+Each connection will be then authenticated by the rusttls server instance.
+
+Once authenticated, the authorisation process will occur as listed below.
+
+
+#### Security
+
+The rustls crate will be utilised as for all cryptogaphic functionality.
+- Since rustls has *NOT* had a security audit yet, this code should definitely *NOT* be used in production.
+- I found rustls to have a better API and docs (from my research) than native-tls.
+
+The x509-parser crate will be used for public key data extraction from client certificates provided to the server.
+
+TLS v1.3 with ECDHE will be used exclusively as it newer, has better performance on initial connection and gives perfect forward secrecy out of the box on all cipher suites.
+
+#### Cipher Suites
+
+The following cipher suites will be:
+    - TLS_AES_256_GCM_SHA384
+    - TLS_AES_128_GCM_SHA256
+
+These cipher suites are supported and enabled by Linux, Apple via OpenSSL/BoringSSL, Windows TLS and rustls.
+
+##### Authentication
+A certificate authority will be created via openssl.
+The server and all client private keys will be signed by the CA.
+
+Since both client and server will be able to authenticate their respective chains of trust, authentication can occur.
+
+As all certificates will be precreated - a bash script will be created which will create:
+    - A CA.
+    - Server private key and cert that is signed by the CA.
+    - N clients private key and cert (with client auth extensions) that is signed by the CA.
+    - Each client will have a unique server name identifier.
+    - All private keys (CA, server, clients) will generated using the elliptic curve - secp256k1.
```

Comment:
nit: if you're considering broader compatibility with SChannel (and presumably SecureTransport, as OpenSSL is very much deprecated on macOS) when picking ciphers, you probably also want to avoid Koblitz curves in favor of just NIST P-256 here.